### ISSUE TITLE
Fibers are more flexible(?)

### ISSUE THREAD

**Issue by gunar** [yortus/asyncawait#51]
The readme states:

> node-fibers: This implementation of coroutines is unfortunately limited to Node.js. ES6 generators may be simpler, but fibers are more flexible and support a far broader space of design possibilities. It would be great if ES6 generators were this open and flexible.

and I got intrigued. What kind of things can Fibers do that Coroutines can't?

Thank you!

**Comment by bjouhier** [comment]
Fibers and coroutines are equivalent. The difference is between fibers and generators.

Fibers and coroutines support _deep_ continuations. This means that you can yield at any depth in the call stack and resume there later.

Generators only support _single frame_ continuations. This means that yielding only saves 1 stack frame. This is less powerful and it explains why you need to yield (or await) at all levels when you use generators (or ES7 async/await) to streamline async call graphs.

**Comment by yortus** [comment]
As an example of yielding at any stack depth as mentioned by @bjouhier, with this fiber-based implementation of async/await you can easily combine async and functional programming:

```js
let testFn = async (() => {

    someArray = [...];

    let results = someArray
        .filter(el => await (someAsyncTest())
        .map(el => await (someAsyncMapping());

    return results;
});
```

Note the two `await`s are not directly within the `async` function, they are actually inside the anonymous lambda functions passed to the `filter` and `map` methods. ES2017 async/await and ES6 generators cannot do that. They only support 'shallow' await/yield which must be directly within the body of an async function/generator.

**Comment by gunar** [comment]
So neither ES207 async/await nor generators+[co](/tj/co]) are proper coroutines? Never thought of that.

@yortus that's an amazing README-worthy example of the benefits!

This could get me to switch over to Fibers/asyncawait. Are there any drawbacks really?

Thank you.

**Comment by bjouhier** [comment]
There is also a performance benefit with some code patterns. 

- If you have many layers of calls on top of async APIs, all the intermediate calls are vanilla JS function calls with fibers. So you don't pay the overhead of allocating a generator or a promise at every call and you benefit from all the optims, like inlining, that the JS engine applies to vanilla calls. 
- If, on the other hand, you only have a thin layer of logic on top of the async APIs, fibers are less efficient because of the `Fiber.yield` calls and because each fiber allocates a stack (lazily, with `mmap`). 

The global balance may vary but with our application fibers are a clear winner.

I had not seen the _Downsides of Fibers_ SO entry before. The first answer is a nice piece of disinformation 👎 :

- Unfamiliarity: coding on top of fiberized APIs feels just like coding on top of sync APIs. Sound pretty familiar, doesn't it?
- Fibers does not work on windows. See https://github.com/laverdet/node-fibers/pull/67. I've been using fibers on windows (with streamline.js) since then mid 2012. Always worked like a charm.
- Browser incompatible: true
- More difficult debugging: nothing can be more wrong than this. As calls are _vanilla_ JS calls, the debugging experience is awesome. Stepping through the code is just like stepping through sync code. And stack traces just work. No need for long-stack-trace hacks.

Coroutines are very popular in go-land but have always been challenged in js-land.

**Comment by gunar** [comment]
That SO answer is at least one year old so yeah.
But debugging is even better than when using `co`? That's nice!

**Comment by bjouhier** [comment]
I haven't done debugging with `co` but I've written a library called `galaxy`, which is similar to `co`. Debugging experience is not great because you have to step through the library at every call. Also there is no API to find out where a generator has been suspended. So it is impossible to generate a precise stack trace of async call chains that go through generators. 

For an (abandoned) attempt of providing the missing API, see https://github.com/bjouhier/galaxy-stack.

**Comment by gunar** [comment]
> because you have to step through the library at every call

So you're saying this doesn't happen with Fibers? Dang I'll just stop talking and give this a try.

**Comment by bjouhier** [comment]
> > because you have to step through the library at every call

> So you're saying this doesn't happen with Fibers? Dang I'll just stop talking and give this a try.

Depends how you write your code. Let's take the example above:

```js
let testFn = async (() => {
    someArray = [...];
    return someArray
        .filter(el => await (someAsyncTest())
        .map(el => await (someAsyncMapping());
});
```
If you write it this way, you will have to call `testFn` as `await testFn()` so you will go through the library at every call.

But, if instead you write:
```js
let testFn = () => {
    someArray = [...];
    return someArray
        .filter(el => await (someAsyncTest())
        .map(el => await (someAsyncMapping());
};
```
Then you can call it as `testFn()` and you can write things like:
``` js
function foo() { return testFn(); }
function bar() { return foo(); }
function zoo() { return bar(); }
```
You don't need wrappers in these calls. You just need `await` whenever your code calls async node APIs, and `async` whenever your code is being called from node APIs. Typical example is:
```js
http.createServer(function(request, response) {
  async(zoo)().catch(err => console.error(err.stack));
}.listen(port);
```

**Comment by bjouhier** [comment]
I forgot to mention parallelizing. If you write all your code as described above, all I/O operations will be serialized and you won't get the parallelizing benefits that node naturally gives you. 

But parallelizing is easy:
```js
const [p1, p2] = [async(zoo)(),  async(zoo)()];
// here the two zoo calls are running concurrently (still single-threaded, but interleaved)
doSomethingElse();
// now we can wait on the two results
const [r1, r2] = [await(p1), await(p2)];
```
Welcome to the wonderful world of coroutines 😄 !

**Comment by gunar** [comment]
Oh man... I finally realized the need for this and it sooo sucks to be stuck on this side :-(

Thank you for enlightening me.
I'll be looking more into `asyncawait`.
I sure wish there was a way to "transpile" `asyncawait` code for the browser.

**Comment by gunar** [comment]
In this example, is `map` blocked by finishing `filter` first?
It must be since this is native to javascript.
But if I were to have that, what name would it have? I'm guess the only way out is Reactive Programming which I am not thrilled by.
```js
let testFn = () => {
    someArray = [...];
    return someArray
        .filter(el => await (someAsyncTest())
        .map(el => await (someAsyncMapping());
};
```

**Comment by bjouhier** [comment]
Yes, `map` is blocked by `filter`.

Not great if the different steps perform I/O because `map` will only start after `filter` has completed all its I/O. It would be better to pipeline the steps.

There are libraries that help with this. My own: https://github.com/Sage/f-streams. At the cross-roads between reactive programming and node.js streams.

**Comment by mikermcneil** [comment]
Great discussion y'all.  Just as a side note, seems like .map() etc could expect an async function for its iteratee.  Imagine:

```js
someArray = await async2.mapSeries(someArray, async (item)=>{
  //...
  return item;
});
```

Or to process each item simultaneously:
```
someArray = await async2.map(someArray, async (item)=>{
  //...
  return item;
});
```

**Comment by mikermcneil** [comment]
I do wish that es8 included a restriction such that invoking an async function would _always_ mandate a keyword prefix. That way, you'd be sure that intent was captured and that teams new to Node.js weren't writing fire-and-forget-code by mistake.

Since async functions are new to JS, we had an opportunity to force userland code to be more clear without breaking backwards compatibility.  For example, we might have had the JS interpreter demand one of the following two syntaxes:

- `var result = await foo()`
- `var promise = begin foo()`


And thus just attempting to call `foo()` normally would throw an error (because foo() is an async function, and it's important to make intent clear).

Currently in JavaScript, calling foo() naively (_unpreceded_ by the special `await` keyword) causes the latter fire-and-forget usage to be assumed.  But unfortunately, that usage is much less common in everyday apps, and also conceptually unexpected for someone new to asynchronous code.

Oh well. It's still a hell of a lot easier to remind someone that they have to use "await" than it is to try and explain how to do asynchronous if statements with self-calling functions that accept callbacks; or the nuances of your favorite promise library; or the pros and cons of external named functions versus inlining asynchronous callbacks.

So I'll take what I can get ☺️

**Comment by dtipson** [comment]
Having to call begin would feel awful similar to, but less powerful than, functional & lazy Futures and .fork...

**Comment by cztomsik** [comment]
Yep, it's impossible to do lazy loading with coroutines but it is possible to do with fibers. Fibers enable true OOP (implementation hiding) and so they are in my opinion superior to (current implementation of) coroutines.

It would be great, if await was allowed in regular functions which would effectively block current coroutine (and throw exception if it is not run in any). I'd love to file proposal to tc39 but I have zero experience with this.

**Comment by bjouhier** [comment]
@erobwen. So true. 

@cztomsik. I'd rather say: "it is impossible to do lazy loading with _stackless_ coroutines". Coroutines come in 2 flavors: stackful (fibers) and stackless (async/await).
